partialmessages/partialmsgs.go

```diff
@@ -0,0 +1,392 @@
+	AvailableParts() ([]byte, error)
+}
+
+type partialMessagePeerState struct {
+	wants    []byte
+	has      []byte
+	sentWant []byte
+	sentHas  []byte
+}
+
+func (ps *partialMessagePeerState) IsZero() bool {
+	return len(ps.wants) == 0 && len(ps.has) == 0 && len(ps.sentHas) == 0 && len(ps.sentWant) == 0
+}
+
+type partialMessageStatePerTopicGroup struct {
+	peerState map[peer.ID]*partialMessagePeerState
+	lastHave  []byte
+	groupTTL  int
+}
+
+func newPartialMessageStatePerTopicGroup(groupTTL int) *partialMessageStatePerTopicGroup {
+	return &partialMessageStatePerTopicGroup{
+		peerState: make(map[peer.ID]*partialMessagePeerState),
+		groupTTL:  max(groupTTL, minGroupTTL),
+	}
+}
+
+func (s *partialMessageStatePerTopicGroup) clearPeerWants(peerID peer.ID) {
+	if peerState, ok := s.peerState[peerID]; ok {
+		if len(peerState.wants) > 0 {
+			peerState.wants = nil
+			if peerState.IsZero() {
+				delete(s.peerState, peerID)
+			}
+		}
+	}
+}
+
+type PartialMessageExtension struct {
+	Logger *slog.Logger
+
+	// OnIncomingRPC is called whenever we receive an encoded
+	// partial message from a peer. This func MUST be fast and non-blocking.
+	// If you need to do slow work, dispatch the work to your own goroutine.
+	OnIncomingRPC func(
+		from peer.ID,
+		topic string,
+		groupID []byte,
+		iwant []byte,
+		ihave []byte,
+		partialMessageBytes []byte,
+	)
+
+	// ValidateRequestMetadata should be a fast function that performs some
+	// basic sanity checks on incoming RPC. For example:
+	//   - Is this a known topic?
+	//   - Is the groupID well formed per application semantics?
+	//   - If this is a PartialIHAVE/PartialIWant, is the request metadata within
+	//     expected bounds?
+	ValidateRPC func(from peer.ID, rpc *pb.PartialMessagesExtension) error
+
+	// GroupTTLByHeatbeat is how many heartbeats we store Group state for.
+	GroupTTLByHeatbeat int
+
+	statePerTopicPerGroup map[string]map[string]*partialMessageStatePerTopicGroup
+
+	router Router
+}
+
+type PublishOptions struct {
+	// PublishToPeers limits the publishing to only the specified peers.
+	// If nil, will use the topic's mesh peers.
+	PublishToPeers []peer.ID
+	// EagerPush is data that will be eagerly pushed to peers in a PartialMessage
+	EagerPush []byte
+}
+
+type Router interface {
+	SendRPC(p peer.ID, r *pb.PartialMessagesExtension, urgent bool)
+	MeshPeers(topic string) iter.Seq[peer.ID]
+}
+
+func (e *PartialMessageExtension) groupState(topic string, groupID []byte) (*partialMessageStatePerTopicGroup, error) {
+	if e.statePerTopicPerGroup == nil {
+		e.statePerTopicPerGroup = make(map[string]map[string]*partialMessageStatePerTopicGroup)
+	}
+	statePerTopic, ok := e.statePerTopicPerGroup[topic]
+	if !ok {
+		statePerTopic = make(map[string]*partialMessageStatePerTopicGroup)
+		e.statePerTopicPerGroup[topic] = statePerTopic
+	}
+	state, ok := statePerTopic[string(groupID)]
+	if !ok {
+		state = newPartialMessageStatePerTopicGroup(max(e.GroupTTLByHeatbeat, minGroupTTL))
```

Comment:
max is already done in `newPartialMessageStatePerTopicGroup`
```suggestion
		state = newPartialMessageStatePerTopicGroup(e.GroupTTLByHeatbeat)
```